clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,340 @@
+     * @param max       The maximum number of examples to look for.  If we find more than
+     *                  this, we will append '...' to the string and return.
+     * @param toFind    The partitions to look for.
+     * @param toSearch  The set of partitions to search.
+     * @return          An empty string if all partitions were found; some of the missing ones
+     *                  in string form, if not.
+     */
+    static String findMissing(int max, Set<TopicPartition> toFind, Set<TopicPartition> toSearch) {
+        StringBuilder bld = new StringBuilder();
+        String prefix = "";
+        int remaining = max;
+        for (TopicPartition part : toFind) {
+            if (!toSearch.contains(part)) {
+                bld.append(prefix);
+                bld.append(part);
+                prefix = ", ";
+                remaining--;
+                if (remaining < 0) {
+                    bld.append("...");
+                    return bld.toString();
+                }
+            }
+        }
+        return bld.toString();
+    }
+
+    /**
+     * Verify that a full fetch response contains all the partitions in the fetch session.
+     *
+     * @param response  The response.
+     * @return          True if the full fetch response partitions are valid.
+     */
+    private boolean verifyFullFetchResponseParts(FetchResponse response) {
+        String missing = findMissing(MAX_MISSING_TO_LOG, response.responseData().keySet(),
+            sessionParts.keySet());
+        if (!missing.isEmpty()) {
+            log.warn("Session handler for node {} received a full fetch response " +
+                "that contained unexpected partition(s) {}", node, missing);
+            return false;
+        }
+        missing = findMissing(MAX_MISSING_TO_LOG, sessionParts.keySet(),
+            response.responseData().keySet());
+        if (!missing.isEmpty()) {
+            log.warn("Session handler for node {} received a full fetch response " +
+                "that did not contain expected partition(s) {}", node, missing);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Verify that the partitions in an incremental fetch response are contained in the session.
+     *
+     * @param response  The response.
+     * @return          True if the incremental fetch response partitions are valid.
+     */
+    private boolean verifyIncrementalFetchResponseParts(FetchResponse response) {
+        String missing = findMissing(MAX_MISSING_TO_LOG, response.responseData().keySet(),
+            sessionParts.keySet());
+        if (!missing.isEmpty()) {
+            log.warn("Session handler for node {} received an incremental fetch response " +
+                "that contained unexpected partition(s) {}", node, missing);
+            return false;
+        }
+        return true;
+    }
+
+    public boolean handleResponse(FetchResponse response) {
+        if (response.error() == Errors.NONE) {
```

Comment:
I was expecting to see some logic to remove a partition from the session following a `NOT_LEADER` error. Maybe I'm missing it somewhere?